### ISSUE TITLE
heap-buffer-overflow w/ASAN

### ISSUE THREAD

**Issue by bill-torpey** [cascadium/OpenMAMA-omnm#12]
Hi Frank:

I'm running into some issues with omnm that appear to be related to large fields/messages, and I've gotten a stack trace using clang's Address Sanitizer that may help provide a clue.  If you can take a moment to look this over, perhaps the light-bulb will go off -- I haven't learned the code well enough to identify the problem yet:

```
==30103==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6110000305bc at pc 0x7f32da168a93 bp 0x7f32d2f12360 sp 0x7f32d2f12358
READ of size 1 at 0x6110000305bc thread T8
    #0 0x7f32da168a92 in omnmmsgPayloadIter_next /home/btorpey/work/OpenMAMA-omnm/nyfix/src/Iterator.cpp:105:9
    #1 0x7f32da16ab80 in OmnmPayloadImpl::findFieldInBuffer(char const*, unsigned short, omnmFieldImpl&) /home/btorpey/work/OpenMAMA-omnm/nyfix/src/Payload.cpp:389:54
    #2 0x7f32da17d8ea in mama_status OmnmPayloadImpl::getFieldValueAsCopy<int>(mamaFieldType_, char const*, unsigned short, int*) /home/btorpey/work/OpenMAMA-omnm/nyfix/src/./Payload.h:127:30
    #3 0x7f32da179945 in omnmmsgPayload_getI32 /home/btorpey/work/OpenMAMA-omnm/nyfix/src/Payload.cpp:2119:39
    #4 0x7f32dfb3c3ec in mamaMsg_getI32 /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/msg.c:1699:12
    #5 0x7f32e182fc6e in Transact::MamaMessageImpl::getOrigTransId(Transact::MXTransId&) const /home/btorpey/work/transact/openmama/src/common/Middleware/MamaAdapter/MamaMessageImpl.cpp:1724:25
    #6 0x5b50fa in TransactX::TxnReq::setMsg(Transact::IMessage*) /home/btorpey/work/transact/openmama/src/server/txnpd/src/TxnReq.cpp:116:17
    #7 0x5de9ad in TransactX::TxnReceiver::onMessage(char const*, Transact::IMessage*) /home/btorpey/work/transact/openmama/src/server/txnpd/src/TxnReceiver.cpp:1352:17
    #8 0x7f32e180c21b in Transact::MamaListenerImpl::onMsg(mamaSubscriptionImpl_*, mamaMsgImpl_*, void*, void*) /home/btorpey/work/transact/openmama/src/common/Middleware/MamaAdapter/MamaListenerImpl.cpp:154:36
    #9 0x7f32df898ef6 in mamaEnvSubscription_onMsgBasic /home/btorpey/work/transact/openmama/src/common/Middleware/MamaAdapter/mme/mamaEnvSubscription.c:308
    #10 0x7f32dfb9bb31 in mamaSubscription_forwardMsg /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/subscription.c:1420:5
    #11 0x7f32dfba0887 in mamaSubscription_processMsg /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/subscription.c:2312:9
    #12 0x7f32dbc70a21 in zmqBridgeMamaTransportImpl_queueCallback /home/btorpey/work/OpenMAMA-zmq/nyfix/src/transport.c:808:16
    #13 0x7f32dfbf51dd in wombatQueue_dispatchInt /shared/work/OpenMAMA/nyfix/common/c_cpp/src/c/queue.c:319:9
    #14 0x7f32dfbf5233 in wombatQueue_timedDispatch /shared/work/OpenMAMA/nyfix/common/c_cpp/src/c/queue.c:335:12
    #15 0x7f32dbc624f6 in zmqBridgeMamaQueue_dispatch /home/btorpey/work/OpenMAMA-zmq/nyfix/src/queue.c:253:16
    #16 0x7f32dfb83f95 in mamaQueue_dispatch /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/queue.c:825:12
    #17 0x7f32dfb85b58 in dispatchThreadProc /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/queue.c:1264:30
    #18 0x33c0607aa0 in start_thread (/lib64/libpthread.so.0+0x33c0607aa0)
    #19 0x33bfee8bcc in clone (/lib64/libc.so.6+0x33bfee8bcc)

0x6110000305bc is located 0 bytes to the right of 252-byte region [0x6110000304c0,0x6110000305bc)
allocated by thread T8 here:
    #0 0x4ec29d in realloc /shared/buildtest/clang/trunk/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:77
    #1 0x7f32dbc76c92 in allocateBufferMemory /shared/work/OpenMAMA/nyfix/common/c_cpp/src/c/memnode.c:109:18
    #2 0x7f32da16d0cb in omnmmsgPayload_unSerialize /home/btorpey/work/OpenMAMA-omnm/nyfix/src/Payload.cpp:900:14
    #3 0x7f32da16e4b2 in omnmmsgPayload_createFromByteBuffer /home/btorpey/work/OpenMAMA-omnm/nyfix/src/Payload.cpp:956:12
    #4 0x7f32dfb3adf9 in mamaMsgImpl_setMsgBuffer /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/msg.c:600:12
    #5 0x7f32dbc5fb4a in zmqBridgeMamaMsgImpl_deserialize /home/btorpey/work/OpenMAMA-zmq/nyfix/src/msg.c:647:13
    #6 0x7f32dbc7090d in zmqBridgeMamaTransportImpl_queueCallback /home/btorpey/work/OpenMAMA-zmq/nyfix/src/transport.c:802:13
    #7 0x7f32dfbf51dd in wombatQueue_dispatchInt /shared/work/OpenMAMA/nyfix/common/c_cpp/src/c/queue.c:319:9
    #8 0x7f32dfbf5233 in wombatQueue_timedDispatch /shared/work/OpenMAMA/nyfix/common/c_cpp/src/c/queue.c:335:12
    #9 0x7f32dbc624f6 in zmqBridgeMamaQueue_dispatch /home/btorpey/work/OpenMAMA-zmq/nyfix/src/queue.c:253:16
    #10 0x7f32dfb83f95 in mamaQueue_dispatch /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/queue.c:825:12
    #11 0x7f32dfb85b58 in dispatchThreadProc /shared/work/OpenMAMA/nyfix/mama/c_cpp/src/c/queue.c:1264:30
    #12 0x33c0607aa0 in start_thread (/lib64/libpthread.so.0+0x33c0607aa0)
    #13 0x33bfee8bcc in clone (/lib64/libc.so.6+0x33bfee8bcc)
```

On a related topic, are there any built-in limitations on field or message size using omnm?  I've been testing assuming that any limitations are <= those from Wombat (i.e., max size of an opaque data type is 0xefff, or 61439 bytes).

Thanks!

Bill

**Comment by fquinner** [comment]
The only limit i can think of is max uint32 for field size in bytes and uint64 max for message size on a 64 bit system and I can't see you hitting that ;). It's possible there is a bug somewhere in there though.

It looks like it is trying to read past the end of allocated memory for the message, which would suggest that `omnmmsgPayloadIter_hasNext` is broken for your use case. That method it depends on `mPayloadBuffer`  and `mPayloadBufferTail` so I'm thinking that somewhere along the line, `mPayloadBuffer` is being realloced (since 252 > default value of 200), and `mPayloadBufferTail` hasn't been updated at the same time. Could you share the code which creates the message in question so I can try a recreation?

**Comment by bill-torpey** [comment]
It does appear to be related to reallocating the buffer -- I changed DEFAULT_PAYLOAD_SIZE to 1024 and I'm still seeing the error, but at a later point in the test cycle where we're deliberately testing (very) large msgs.

There's really nothing special about the message in the original stack trace -- it just happened to go over the 200 byte initial allocation (in omnmmsgPayload_unSerialize). 

I'll bang on it some more, but in the meantime I think all you need to do is create a message with a payload size > DEFAULT_PAYLOAD_SIZE (or reduce DEFAULT_PAYLOAD_SIZE to make it happen sooner), and then call a get method.

**Comment by bill-torpey** [comment]
Been tracing through in the debugger, and I run into this code (in updateField) that looks suspicious to say the least:

~~~ c++
        // If the field has changed in size (increased or decreased), we need to move memory
        if (delta != 0)
        {
            // This will point to the target location of the first field after inserted field
            uint8_t* origin = mPayloadBuffer + nextByteOffset;
            // This will correspond to the number of remaining bytes after the updated field
            uint32_t size = mPayloadBufferTail - nextByteOffset;
            // Finally move the memory across
            memmove ((void*)(origin + delta), origin, size);
        }
        mPayloadBufferTail += delta;
    }
~~~

1. The calculation of origin assumes that the updated field will be appended to the end of (possibly new) buffer.
 - This is not necessary if the updated field is the same size or smaller (not relevant in this case, but just mentioning it).
 - This will leave a "hole" in the buffer where the old field was, but I don't see any code that accounts for that by setting the old contents to some value that would cause an iterator to skip over the hole.  Can you point that out?   
2. Given #1, I don't understand the point of using memmove -- there is no possibility that the source and destination will overlap.
 - I also don't understand what the memmove is doing in the first place -- doesn't it need to move the *entire* field (incl. type, fid, etc.) to the new location?  It doesn't seem to be doing that...

Maybe I'm just not understanding how this is supposed to work -- if the above is incorrect, please let me know.  Thanks!

**Comment by fquinner** [comment]
Yeah that logic looks like it assumes a signed value for `nextByteOffset` on phone now will hopefully get better look later on

**Comment by bill-torpey** [comment]
Thanks!  I'm happy to do more testing on my end, but I want to try to understand the code first ;-)

**Comment by bill-torpey** [comment]
Not sure if this helps, but it may be relevant to the discussion of "holes" in the buffer...

When I run a simple test pgm that populates a blob field with sizes= 100, 256 and 1K, this is what I see from mamaMsg_asString:

```
9/14 17:08:15.007703|MamaAdapterTest|9617-140160855399680|INFO(0,0) Publisher 0x603000006970 sending msg #1 with size(218), contents({field1=1234,field2=4321,field3=value3,field4=value4,test=value4,blobType=100,msgNo=1})|blobTest.cpp(55)
9/14 17:08:15.023676|MamaAdapterTest|9617-140160648730368|INFO(0,0) Listener 0x6030000069d0 received msg #1 with size(241), contents({field1=1234,field2=4321,field3=value3,field4=value4,test=value4,blobType=100,msgNo=1,__sbj=topic,=Unknown,=Unknown})|blobTest.cpp(85)
9/14 17:08:15.024525|MamaAdapterTest|9617-140160648730368|INFO(0,0) Retrieved blob field w/size=0x64|blobTest.cpp(92)
9/14 17:08:16.010972|MamaAdapterTest|9617-140160855399680|INFO(0,0) Publisher 0x603000006970 sending msg #2 with size(393), contents({field1=1234,field2=4321,field3=value3,field4=value4,test=value4,=Unknown,blobType=256,msgNo=2,__sbj=topic})|blobTest.cpp(55)
9/14 17:08:16.012776|MamaAdapterTest|9617-140160648730368|INFO(0,0) Listener 0x6030000069d0 received msg #2 with size(401), contents({field1=1234,field2=4321,field3=value3,field4=value4,test=value4,=Unknown,blobType=256,msgNo=2,__sbj=topic,=Unknown,=Unknown})|blobTest.cpp(85)
9/14 17:08:16.013567|MamaAdapterTest|9617-140160648730368|INFO(0,0) Retrieved blob field w/size=0x100|blobTest.cpp(92)
9/14 17:08:17.012137|MamaAdapterTest|9617-140160855399680|INFO(0,0) Publisher 0x603000006970 sending msg #3 with size(1164), contents({field1=1234,field2=4321,field3=value3,field4=value4,test=value4,=Unknown,=Unknown,blobType=1k,msgNo=3,__sbj=topic})|blobTest.cpp(55)
9/14 17:08:17.013982|MamaAdapterTest|9617-140160648730368|INFO(0,0) Listener 0x6030000069d0 received msg #3 with size(1172), contents({field1=1234,field2=4321,field3=value3,field4=value4,test=value4,=Unknown,=Unknown,blobType=1k,msgNo=3,__sbj=topic,=Unknown,=Unknown})|blobTest.cpp(85)
9/14 17:08:17.014787|MamaAdapterTest|9617-140160648730368|INFO(0,0) Retrieved blob field w/size=0x400|blobTest.cpp(92)
```

The relevant bits:
1.  For all messages the value of the blob field (test) is printed as the value of the "value4" field.
2.  Once buffer exceeds DEFAULT_PAYLOAD_SIZE,  there are spurious "=Unknown" entries printed.
3. For all messages that come off the wire, there are spurious "=Unknown" entries printed.

**Comment by fquinner** [comment]
Just checked there and that logic actually looks OK and can be unsigned since it will always be positive (it represents the number of bytes offset from the payload's root pointer where the *next* field after this one starts. It's specifically designed to fill these holes and there is even a unit test which checks for shrinking and growing buffers.

Going through your queries on the code:

> The calculation of origin assumes that the updated field will be appended to the end of (possibly new) buffer.
> This is not necessary if the updated field is the same size or smaller (not relevant in this case, but just mentioning it).

That's what the delta variable is for - tracks how many bytes it has increased or decreased by. That code is only hit if a change in size has occurred.

> This will leave a "hole" in the buffer where the old field was, but I don't see any code that accounts for that by setting the old contents to some value that would cause an iterator to skip over the hole. Can you point that out?

Each field is part of a contiguous buffer, so the logic here is that if necessary, the entire buffer (containing all fields) is realloc'd. After the realloc there should be no holes - just some undefined memory at the end of the memory block that we're about to fill in the event of an increase in size. If the field value has gotten *smaller*, the memmove later fills that hole

> Given #1, I don't understand the point of using memmove -- there is no possibility that the source and destination will overlap.
> I also don't understand what the memmove is doing in the first place -- doesn't it need to move the entire field (incl. type, fid, etc.) to the new location? It doesn't seem to be doing that...

The memmove doesn't move the field itself - it moves the *rest* of the contiguous block containing all subsequent fields. There is code above which re-populates `field.mData` and `field.mName` after the alloc based on values based on byte offsets from the new buffer location to be safe after potential realloc.

So i'm not quite sure where the problem actually lives. I tried making a version of https://github.com/fquinner/OpenMAMA-omnm/blob/master/src/UnitTests.cpp#L265 which forced a realloc but it looked ok. Any chance you could share `blobTest.cpp`?

**Comment by bill-torpey** [comment]
Thanks Frank!

The part I didn't get was that you're actually reallocating the data *in place* and shifting everything following the new data to the right to make room.  I'll take a look at the code again with that in mind (no wonder I was confused ;-) and see if can figure out where things are going wrong.

I've tried the same as you with the unit tests, and similarly no joy.

I'd love to share blobTest with you, but it's built on our custom framework, which (a) has a bunch of dependencies that you definitely don't want, and (b) I'd get fired :-(  In any event, that's not what ASAN was complaining about.

Let me bang on this for a while and I'll get back to you when I learn something.

Thanks again!

**Comment by fquinner** [comment]
Hi Bill did you ever get to the bottom of this?

**Comment by bill-torpey** [comment]
The original problem was resolved, either by https://github.com/nyfix/OpenMAMA-omnm/commit/a6c7d4fd5d434ae6a2533bc0f7062e08cabe9f94 or when we changed the payload code as per https://github.com/cascadium/OpenMAMA-omnm/pull/15.

I believe the original problem still exists -- if a message needs to be reallocated, the code has an "off-by-one" error or some-such.

Note that this is ONLY visible w/ASAN (and prob. valgrind too) -- it does NOT necessarily cause a SEGV. 

https://github.com/cascadium/OpenMAMA-omnm/pull/17/files has relevant script changes to enable ASAN.
